les/vflux/server/clientpool.go

```diff
@@ -0,0 +1,375 @@
+	"github.com/ethereum/go-ethereum/les/vflux"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/metrics"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethereum/go-ethereum/p2p/nodestate"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+var (
+	serverSetup = &nodestate.Setup{}
+	clientField = serverSetup.NewField("client", reflect.TypeOf(clientPeerInstance{}))
+	btSetup     = newBalanceTrackerSetup(serverSetup)
+	ppSetup     = newPriorityPoolSetup(serverSetup)
+)
+
+var (
+	ErrNotConnected    = errors.New("client not connected")
+	ErrNoPriority      = errors.New("priority too low to raise capacity")
+	ErrCantFindMaximum = errors.New("Unable to find maximum allowed capacity")
+)
+
+func init() {
+	btSetup.connect(clientField, ppSetup.capacityField)
+	ppSetup.connect(btSetup.balanceField, btSetup.updateFlag) // nodeBalance implements nodePriority
+}
+
+// ClientPool implements a client database that assigns a priority to each client
+// based on a positive and negative balance. Positive balance is externally assigned
+// to prioritized clients and is decreased with connection time and processed
+// requests (unless the price factors are zero). If the positive balance is zero
+// then negative balance is accumulated.
+//
+// Balance tracking and priority calculation for connected clients is done by
+// balanceTracker. activeQueue ensures that clients with the lowest positive or
+// highest negative balance get evicted when the total capacity allowance is full
+// and new clients with a better balance want to connect.
+//
+// Already connected nodes receive a small bias in their favor in order to avoid
+// accepting and instantly kicking out clients. In theory, we try to ensure that
+// each client can have several minutes of connection time.
+//
+// Balances of disconnected clients are stored in nodeDB including positive balance
+// and negative banalce. Boeth positive balance and negative balance will decrease
+// exponentially. If the balance is low enough, then the record will be dropped.
+type ClientPool struct {
+	*priorityPool
+	*balanceTracker
+	clock  mclock.Clock
+	closed bool
+	ns     *nodestate.NodeStateMachine
+	synced func() bool
+
+	lock                                 sync.RWMutex
+	defaultPosFactors, defaultNegFactors PriceFactors
+	connectedBias                        time.Duration
+
+	minCap     uint64      // the minimal capacity value allowed for any client
+	capReqNode *enode.Node // node that is requesting capacity change; only used inside NSM operation
+
+	capacityQueryZeroMeter, capacityQueryNonZeroMeter metrics.Meter
+}
+
+// clientPeer represents a peer in the client pool. None of the callbacks should block.
+type clientPeer interface {
+	Node() *enode.Node
+	FreeClientId() string                         // unique id for non-priority clients (typically a prefix of the network address)
+	InactiveTimeout() time.Duration               // disconnection timeout for inactive non-priority peers
```

Comment:
I would prefer the name `InactiveAllowance`?

Reply:
done